scylla-server-rust/src/controller/data_controller.rs

```diff
@@ -0,0 +1,42 @@
+use axum::{
+    extract::{Path, State},
+    Json,
+};
+use serde::Serialize;
+
+use crate::{
+    error::ScyllaError,
+    prisma::{self},
+    services::data_service,
+    Database,
+};
+
+#[derive(Serialize, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub struct DataSend {
+    pub time: i64,
+    pub values: Vec<String>,
+}
+
+// convert the prisma type to the client type for JSON encoding
+impl From<&prisma::data::Data> for DataSend {
+    fn from(value: &prisma::data::Data) -> Self {
+        DataSend {
+            values: value.values.iter().map(|f| f.to_string()).collect(),
+            time: value.time.timestamp_millis(),
+        }
+    }
+}
+
+pub async fn get_data(
+    State(db): State<Database>,
+    Path((data_type_name, run_id)): Path<(String, i32)>,
+) -> Result<Json<Vec<DataSend>>, ScyllaError> {
+    let data = data_service::get_data(&db, data_type_name, run_id).await?;
```

Comment:
if this errors, will it automatically return the Error? Im confused why this is returning a result? this should be a void function that sends the data via whatever http interface we are gonna use. And we should handle errors either from here or with a custom error handler.

Reply:
Axum has two error handlers, of which I implemented the infallible one so far.  This basically means a controller cannot and will not panic.  If it does panic in the controller it moves on to the fallible error handler, which i have not done yet.  Basically returning a result means the Ok() value is the payload sent to the client and error is sent to the error handler system in `error.rs`.  The `IntoResponse` is called therefore processing the error to send to the client.  This has multiple benefits, including allowing for a very effortless `?` operator for any result we have to `unwrap()` usually.   It also allows for custom error handling to be statically typed, which increases readability and safety assurances.